packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts

```diff
@@ -80,37 +122,50 @@ export class TreeNodeKernel implements Listenable<TreeChangeEvents> {
 		treeNodeToKernel.set(node, this);
 	}
 
+	/**
+	 * Transition from {@link Unhydrated} to hydrated.
+	 * @remarks
+	 * Happens at most once for any given node.
+	 */
 	public hydrate(anchorNode: AnchorNode): void {
-		const offChildrenChanged = anchorNode.on("childrenChangedAfterBatch", () => {
-			this.#events.emit("nodeChanged");
-		});
-
-		const offSubtreeChanged = anchorNode.on("subtreeChangedAfterBatch", () => {
-			this.#events.emit("treeChanged");
-		});
-
-		const offAfterDestroy = anchorNode.on("afterDestroy", () => this.dispose());
+		assert(!this.disposed, "cannot use a disposed node");
 
 		this.#hydrated = {
 			anchorNode,
-			offAnchorNode: () => {
-				offChildrenChanged();
-				offSubtreeChanged();
-				offAfterDestroy();
-			},
+			offAnchorNode: new Set([
+				anchorNode.on("afterDestroy", () => this.dispose()),
+				// TODO: this should be triggered on change even for unhydrated nodes.
+				anchorNode.on("childrenChanging", () => {
+					this.generationNumber += 1;
+				}),
+			]),
 		};
-	}
 
-	public dehydrate(): void {
-		this.#hydrated?.offAnchorNode?.();
-		this.#hydrated = undefined;
+		// If needed, register forwarding emitters for events from before hydration
+		if (this.#preHydrationEvents !== undefined) {
```

Comment:
Nit: early return on `this.#preHydrationEvents === undefined` would reduce nesting some here.

Reply:
I prefer this as is, since if someone needs to add more logic to this function in the future, this way they can just add it at the top level in what ever order it makes sense, whereas if we make your proposed change, if they add it to the end of the function, it won't run if there were prehydration events, which is almost certainly unrelated to what ever logic they might want to add.